In [19]:
import pandas as pd
import numpy as np

Let's read in our dataset

In [9]:
df = pd.read_csv("./FoodBalanceSheets_E_Africa_NOFLAG.csv", encoding='latin-1')
df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

# **11. What is the total sum of Animal fat produced in 2014 and 2017 respectively?**

In [10]:
df.groupby("Item").sum()

Area Code  Item Code  Element Code      Y2014      Y2015  \
Item                                                                            
Alcohol, Non-Food        38447     752497       1515090    2403.00    2180.00   
Alcoholic Beverages      75016    1649136       2171381  102410.11   98783.72   
Animal Products          18060     397035         90990   11935.65   11811.73   
Animal fats              85279    1882494       2567303  209460.54  200675.72   
Apples and products      69280    1352989       1920418    9499.23   10559.15   
...                        ...        ...           ...        ...        ...   
Vegetables, Other        77779    1518715       2275409  155038.96  158104.08   
Vegetal Products         18060     391905         90990  107145.19  107064.17   
Wheat and products       87073    1639683       2654934  232670.13  234710.51   
Wine                     66206    1319535       1818328    4497.36    4251.81   
Yams                     50798     970905       1525745  200396.96  203151.78   

                         Y2016      Y2017      Y2018  
Item                                                  
Alcohol, Non-Food      2223.00    2348.00    2293.00  
Alcoholic Beverages   96958.75   95581.06   97847.27  
Animal Products       11661.69   11547.65   11578.61  
Animal fats          183314.22  269617.53  269648.27  
Apples and products    9853.26   10198.90    9640.51  
...                        ...        ...        ...  
Vegetables, Other    156218.90  157752.59  163987.21  
Vegetal Products     106792.72  107655.20  107775.39  
Wheat and products   239377.07  240047.62  242645.19  
Wine                   3872.09    4178.02    4039.32  
Yams                 220626.20  229174.59  221272.09  

[119 rows x 8 columns]

According to the groupby function the answers are 209460.54 and 269617.53 for the year 2014 and 2017 respectively.

# **12. What is the mean and standard deviation across the whole dataset for the year 2015 to 3 decimal places?**

In [11]:
df.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

# **13. What is the total number and percentage of missing data in 2016 to 2 decimal places?**

In [18]:
df.isnull().sum()

Area Code          0
Area               0
Item Code          0
Item               0
Element Code       0
Element            0
Unit               0
Y2014           1589
Y2015           1548
Y2016           1535
Y2017           1506
Y2018           1436
dtype: int64

# **14. Which year had the highest correlation with "Element Code"**

In [21]:
df.corr().abs()

Area Code  Item Code  Element Code     Y2014     Y2015  \
Area Code      1.000000   0.005159      0.000209  0.006164  0.005472   
Item Code      0.005159   1.000000      0.024683  0.021722  0.020857   
Element Code   0.000209   0.024683      1.000000  0.024457  0.023889   
Y2014          0.006164   0.021722      0.024457  1.000000  0.994647   
Y2015          0.005472   0.020857      0.023889  0.994647  1.000000   
Y2016          0.005247   0.020109      0.023444  0.996081  0.995739   
Y2017          0.005006   0.021494      0.024254  0.995230  0.988048   
Y2018          0.005665   0.021314      0.024279  0.994872  0.988208   

                 Y2016     Y2017     Y2018  
Area Code     0.005247  0.005006  0.005665  
Item Code     0.020109  0.021494  0.021314  
Element Code  0.023444  0.024254  0.024279  
Y2014         0.996081  0.995230  0.994872  
Y2015         0.995739  0.988048  0.988208  
Y2016         1.000000  0.992785  0.992757  
Y2017         0.992785  1.000000  0.998103  
Y2018         0.992757  0.998103  1.000000

# **15. What year has the highest sum of Import Quantity**

In [29]:
df.groupby("Element").sum()[["Y2014", "Y2015", "Y2016", "Y2017", "Y2018"]].loc["Import Quantity"].sort_values(ascending=False)

Y2017    294559.09
Y2018    287997.09
Y2016    286582.78
Y2014    274144.48
Y2015    267018.46
Name: Import Quantity, dtype: float64

# **16. What is the total number of the sum of Production in 2014?**

In [33]:
df.groupby("Element").sum().loc["Production"]

Area Code         526751.00
Item Code       10450053.00
Element Code    21388191.00
Y2014            1931287.75
Y2015            1947019.39
Y2016            1943537.15
Y2017            2030056.89
Y2018            2075072.89
Name: Production, dtype: float64

# **17. Which of these elements had the highest sum in 2018?**

In [47]:
df[["Y2018", "Element"]].groupby("Element").sum()

Y2018
Element                                           
Domestic supply quantity                2161192.10
Export Quantity                          181594.80
Fat supply quantity (g/capita/day)        10258.69
Feed                                     233489.68
Food                                    1303841.28
Food supply (kcal/capita/day)            455261.00
Food supply quantity (kg/capita/yr)       49056.85
Import Quantity                          287997.09
Losses                                   163902.00
Other uses (non-food)                     91300.97
Processing                               308429.00
Production                              2075072.89
Protein supply quantity (g/capita/day)    11833.56
Residuals                                 34864.00
Seed                                      25263.14
Stock Variation                           20577.91
Total Population - Both sexes           1140605.00
Tourist consumption                          90.00

# **18. Which of these elements had the 3rd lowest sum in 2018**

In [48]:
df[["Y2018", "Element"]].groupby("Element").sum()

Y2018
Element                                           
Domestic supply quantity                2161192.10
Export Quantity                          181594.80
Fat supply quantity (g/capita/day)        10258.69
Feed                                     233489.68
Food                                    1303841.28
Food supply (kcal/capita/day)            455261.00
Food supply quantity (kg/capita/yr)       49056.85
Import Quantity                          287997.09
Losses                                   163902.00
Other uses (non-food)                     91300.97
Processing                               308429.00
Production                              2075072.89
Protein supply quantity (g/capita/day)    11833.56
Residuals                                 34864.00
Seed                                      25263.14
Stock Variation                           20577.91
Total Population - Both sexes           1140605.00
Tourist consumption                          90.00

# **19. What is the total import Quantity in Algeria in 2018?**

In [52]:
df[["Area", "Element", "Y2018"]].groupby(["Element", "Area"]).sum().loc["Import Quantity"]

Y2018
Area                                 
Algeria                      36238.29
Angola                        7769.67
Benin                         7008.24
Botswana                      1908.42
Burkina Faso                  2605.64
Cabo Verde                     499.66
Cameroon                      4188.88
Central African Republic        75.02
Chad                           491.95
Comoros                          2.92
Congo                         1785.27
Côte d'Ivoire                 9271.21
Djibouti                      3784.92
Egypt                        53954.20
Eswatini                       936.53
Ethiopia                      5655.24
Ethiopia PDR                     0.00
Gabon                         1079.75
Gambia                        1212.10
Ghana                         9914.68
Guinea                        4707.95
Guinea-Bissau                  497.72
Kenya                        11188.96
Lesotho                       1016.97
Liberia                       1203.77
Madagascar                    3214.79
Malawi                         520.20
Mali                          1962.06
Mauritania                    3575.74
Mauritius                     1672.39
Morocco                      19448.50
Mozambique                    5552.03
Namibia                       1462.04
Niger                         2390.20
Nigeria                      21923.38
Rwanda                        1734.44
Sao Tome and Principe           84.10
Senegal                       8005.24
Seychelles                      77.71
Sierra Leone                  1659.22
South Africa                 16192.45
Sudan                         9130.11
Sudan (former)                   0.00
Togo                          3625.48
Tunisia                      11030.91
Uganda                        3209.87
United Republic of Tanzania   1863.68
Zambia                         990.75
Zimbabwe                      1673.84

# **20. What is the total number of unique countries in the dataset?**

In [56]:
df["Area"].nunique()

49